### ISSUE TITLE
Pipeline support

### ISSUE THREAD

**Issue by bdpedigo** [microsoft/graspologic#69]
Roll several preprocessing steps (various types of PTR, other transforms, etc), embeddings, and clustering steps into a sklearn Pipeline that will work with randomized parameter search

https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html#sklearn.pipeline.Pipeline

https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html#sklearn.model_selection.RandomizedSearchCV

**Comment by bdpedigo** [comment]
@j1c Since we are going to support `mug2vec`, and @QuasiLegendre @tathey1 and others are interested in chaining graspy functions, I think we should decide how to support pipeline-type things. According to sklearn glossary, [transformers](https://scikit-learn.org/stable/glossary.html#term-transformers) must be [transductive](https://scikit-learn.org/stable/glossary.html#term-transductive). Since we have not implemented true transform methods for many of our embeddings (I don't think they are truly transductive), we aren't supported by `sklearn.pipeline`.  But, after reading the docs, the only reason I see for changing our stuff to have some kind of transform method would be to take advantage of the model selection utilities in sklearn such as `GridSearchCV`. @tathey1 is that something you are interested in doing? 

Another idea is to write our own `graspy.pipeline` function/class that works with our own stuff, I think it would basically just iterate through a list of functions/classes and handle the kwargs, etc.

**Comment by bdpedigo** [comment]
#237 #236

**Comment by j1c** [comment]
So yea I've looked at the issue, and it actually does make sense to implement a true `transform` function for `MASE`. That is, once it learned the common latent position matrix, it can be used to `transform` any other graph(s) with matched vertices. Similar to, say, PCA where you can learn the components from one dataset, and apply the components to different data assuming the features are the same. This is trivial to implement and we should do this.

The problem I see is with `ASE`, `LSE` and `Omni` where doing this out of sample embeddings is hard and not intuitive. Perhaps one solution is to absorb `OutOfSampleASE` to be the transform function for ASE and LSE. I see #237 with a hacky solution to this for `MASE`, but I don't feel comfortable calling that `transform` function since `transform` necessarily means that it can take in data that was not used to fit the model. 

Im also fine with writing our own pipeline.

**Comment by j1c** [comment]
> For an estimator to be usable together with pipeline.Pipeline in any but the last step, it needs to provide a fit or fit_transform function. To be able to evaluate the pipeline on any data but the training set, it also needs to provide a transform function. There are no special requirements for the last step in a pipeline, except that it has a fit function. All fit and fit_transform functions must take arguments X, y, even if y is not used. Similarly, for score to be usable, the last step of the pipeline needs to have a score function that accepts an optional y.

From: https://scikit-learn.org/stable/developers/contributing.html#pipeline-compatibility

I propose that we do not have a `transform` function for `ASE`, `LSE` and `Omni`. I don't really know what is happening in #237.

**Comment by bdpedigo** [comment]
Agreed. I think with `MASE` there could be an argument for the transform function projecting (potentially new) graphs into the space that was `fit`. I talked with @tathey1 and @QuasiLegendre  about this a bit today, and I said they could essentially wrap the `MASE` class or just make changes to the class on a branch if they want to move forward with pipelineing, and possibly create something like `mug2vec` which uses `MASE` instead. @tathey1 @QuasiLegendre can you chime in and make sure we are understanding your plans for this?

**Comment by QuasiLegendre** [comment]
I have changed my code and begin to use different classes to propose these
features, and I need to change the name of those new-made function. For the
mug2vec, I need to read it first. Thanks for your help.

Benjamin Pedigo <notifications@github.com> 于2019年7月26日周五 下午3:45写道：
